### PR TITLE
Readable account trait for stored account meta

### PR DESCRIPTION
**Pull request by jeffwashington**
#### Problem
Remove copies of account and its data during a few code paths.
#### Summary of Changes

Fixes #

### CODE REVIEW

**Comment by jeffwashington** [comment]
note this depends on a few other prs that are outstanding.

File: runtime/src/accounts_db.rs [hunk 1]
```diff
@@ -1137,6 +1137,33 @@ impl solana_frozen_abi::abi_example::AbiExample for AccountsDb {
     }
 }
 
+impl<'a> ReadableAccount for StoredAccountMeta<'a> {
```
Comment (jeffwashington) [inline runtime/src/accounts_db.rs:1140]: @steviez here is the future

File: runtime/src/accounts_db.rs [hunk 2]
```diff
@@ -3691,50 +3724,52 @@ impl AccountsDb {
...
-                }
+                account_info
             })
             .collect()
     }
 
     fn store_accounts_to<
         F: FnMut(Slot, usize) -> Arc<AccountStorageEntry>,
         P: Iterator<Item = u64>,
+        T: ReadableAccount,
```
Comment (carllin) [inline runtime/src/accounts_db.rs:3754]: Will `T` here always be an `AccountSharedData`, or is there something else we have in mind?

Comment (jeffwashington) [reply runtime/src/accounts_db.rs:3754]: comment right above here is an example that isn't AccountSharedData.
https://github.com/solana-labs/solana/pull/16690#discussion_r621561972
Also, serialized accounts from program execution are possible. Or, something that holds Arc<AccountSharedData> internally or holds the contents of AccountSharedData in a single vector allocation from load time (more like a store or serialized for execution).

My plan was to make the T: ReadableAccount or impl ReadableAccount change after the change for Option<&AccountSharedData> I've already changed other code paths to ReadableAccount. That was the plan all along. AccountSharedData is slowly but surely having no more public members and all access is through ReadableAccount.

Comment (jeffwashington) [reply runtime/src/accounts_db.rs:3754]: sorry - this pr wasn't meant for review yet. it is a test bed for several ideas.

Comment (jeffwashington) [reply runtime/src/accounts_db.rs:3754]: steve had just asked where I was going and some 'why'. Feel free to discuss or ask other questions.

**Review by steviez — APPROVED** [review]
Ship it!

**Review by brooksprumo — APPROVED** [review]

File: runtime/src/accounts_db.rs [hunk 3]
```diff
@@ -1964,14 +1982,14 @@ impl AccountsDb {
             let mut write_versions = Vec::with_capacity(alive_accounts.len());
 
             for (_pubkey, alive_account) in alive_accounts.iter() {
-                stored_accounts.push(alive_account.account.clone_account());
+                stored_accounts.push(&alive_account.account);
```
[the remark this answers is not in the document]
Comment (carllin) [reply runtime/src/accounts_db.rs:1985]: whoo awesome!

**Review by carllin — APPROVED** [review]

**Pull request merged**